### ISSUE TITLE
Cron schedule is being duplicated

### ISSUE THREAD

**Issue by udovicic** [magento/magento2#21380]
### Preconditions (*)

1. Create empty Magento project using `composer create-project --repository-url=https://repo.magento.com/ magento/project-community-edition:2.2.7 .`
2. Install with something like following: `php bin/magento setup:install --base-url="http://www.example.com/" --db-host="db" --db-name="..." --db-user="..." --db-password="..." --admin-firstname="Admin" --admin-lastname="Admin" --admin-email="admin@example.com" --admin-user="admin" --admin-password="..." --language="en_US" --currency="EUR" --timezone="Europe/Zagreb" --use-rewrites="1" --backend-frontname="admin"`. This is actually blank installation, but lets document just in case.

### Steps to reproduce (*)

1. php bin/magento cron:run
2. php bin/magento cache:flush
3. (not sure, but i think you need to wait for 1 min)
4. php bin/magento cron:run
5. Execute following on the database: ```SELECT count(*), job_code, scheduled_at FROM `cron_schedule` GROUP BY job_code, scheduled_at ORDER BY `count(*)` DESC```

### Expected result (*)

1. There are no duplicate cron jobs

### Actual result (*)

1. Cron jobs are being duplicated: Same job code at the same time exists multiple times

### Note:
This has been reported few times before, but issue has been closed with note that new one should be opened if it still persists.

**Comment by magento-engcom-team** [comment]
Hi @udovicic. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [x] Summary of the issue
- [ ] Information on your environment
- [x] Steps to reproduce
- [x] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me 2.3-develop instance` - upcoming 2.3.x release

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@udovicic do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [x] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

-  _Next steps are available in case you are a member of **Community Maintainers**._

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by ghost** [comment]
Hi @udovicic this is expected behavior, if you have a job that runs every time like * * * * *, magento create a few rows for this job. this is not publicated rows.
![deepinscreenshot_select-area_20190221121239](https://user-images.githubusercontent.com/40756439/53161112-02546380-35d2-11e9-83cc-c7370f26c9a3.png)

**Comment by udovicic** [comment]
Hi @engcom-backlog-nazar ,

If you repeat steps 1. - 4. few times, more duplicated rows appear, for an example:

count | job_codeJob Code | scheduled_at
--- | --- | ---
2 | indexer_update_all_views | 2019-02-21 10:02:00
2 | sales_send_order_invoice_emails | 2019-02-21 10:02:00
2 | sales_grid_order_invoice_async_insert | 2019-02-21 10:02:00
2 | catalog_product_outdated_price_values_cleanup | 2019-02-21 10:02:00
2 | sales_send_order_creditmemo_emails | 2019-02-21 10:02:00
2 | indexer_reindex_all_invalid | 2019-02-21 10:01:00
2 | outdated_authentication_failures_cleanup | 2019-02-21 10:02:00
2 | sales_grid_order_shipment_async_insert | 2019-02-21 10:02:00
2 | catalog_product_frontend_actions_flush | 2019-02-21 10:02:00
2 | sales_send_order_shipment_emails | 2019-02-21 10:02:00
2 | indexer_reindex_all_invalid | 2019-02-21 10:02:00
2 | sales_grid_order_creditmemo_async_insert | 2019-02-21 10:02:00
2 | magento_newrelicreporting_cron | 2019-02-21 10:02:00
2 | sales_send_order_emails | 2019-02-21 10:02:00
2 | sales_grid_order_async_insert | 2019-02-21 10:02:00
2 | indexer_update_all_views | 2019-02-21 10:01:00
1 | sales_grid_order_creditmemo_async_insert | 2019-02-21 10:22:00
1 | sales_send_order_creditmemo_emails | 2019-02-21 10:16:00
1 | sales_send_order_emails | 2019-02-21 10:04:00
1 | sales_grid_order_async_insert | 2019-02-21 10:04:00
1 | indexer_update_all_views | 2019-02-21 10:07:00
1 | outdated_authentication_failures_cleanup | 2019-02-21 10:12:00
1 | sales_send_order_emails | 2019-02-21 10:20:00
1 | sales_grid_order_async_insert | 2019-02-21 10:20:00
1 | magento_newrelicreporting_cron | 2019-02-21 10:08:00


 Most of them are the cases that you have mentioned above, but one of the rows that also showed for me is:

```xml
<job name="magento_newrelicreporting_cron" instance="Magento\NewRelicReporting\Model\Cron" method="runCron">
    <schedule>*/2 * * * *</schedule>
</job>
```

Just an example to show, but I do not think that one should be duplicated.

**Comment by ghost** [comment]
@udovicic on you screen you have 2 defferent job starting at the same time waths wrong here ?

**Comment by udovicic** [comment]
@engcom-backlog-nazar That is not correct, if you check the SQL from description, you will notice that first column shows number of occurrences. Meaning that there is only one job `magento_newrelicreporting_cron` scheduled for `2019-02-21 10:08:00`, but for `2019-02-21 10:02:00` it is scheduled two times.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-backlog-nazar 
 Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-98358`, `MAGETWO-98359` were created

**Issue Available**: @engcom-backlog-nazar, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by amol2jcommerce** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  amol@twojay.in
    Domain 2jcommerce.in has exceeded the max emails per hour (257/250 (102%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from out-16.smtp.github.com ([192.30.254.199]:34653)
	by mail.2jcommerce.co.uk with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.87)
	(envelope-from <noreply@github.com>)
	id 1gwlp0-0000jx-6w
	for amol@2jcommerce.in; Thu, 21 Feb 2019 10:43:18 +0000
Date: Thu, 21 Feb 2019 02:42:36 -0800
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1550745756;
	bh=yDOWfInhXI4Xm+BigWYUUgxbLRJgmdBDS8krpzce5kk=;
	h=Date:From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=aJuWT062XvMMziWOdW7x4Wy2vBXW01goux2ZM9lmgSMOlft9rJOTojeFLNpCYka9Y
	 /tFrbFY0+Hf4fPs8LwCebcv7hBz8R7Wg8pXT7eDPJYuklaI7X+RwboHAPPudnGSo2D
	 OehVEwGoqeD6XQDTzVnfW1Y4yTuvrMiFauBKw5pM=

**Comment by dipti2jcommerce** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  dipti@twojay.in
    Domain 2jcommerce.in has exceeded the max emails per hour (298/250 (119%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
------ The body of the message is 20056 characters long; only the first
------ 5000 or so are included here.
Received: from out-7.smtp.github.com ([192.30.252.198]:48957)
	by mail.2jcommerce.co.uk with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.87)
	(envelope-from <noreply@github.com>)
	id 1gwojd-000GRF-Cx
	for dipti@2jcommerce.in; Thu, 21 Feb 2019 13:49:57 +0000
Date: Thu, 21 Feb 2019 05:49:16 -0800
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1550756956;
	bh=aC9ZBlyUqZp9sKWKWkLkeVu/ixwsHORkMBLSAarfDkY=;
	h=Date:From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=c6MDBjsPtporaEjXySLTSSWuNJnZxfATJN6zFrn/7tXI6Mn9Jt4Xi3+ZoRtIjG3cG
	 r9iIQeH/7HbQTjYNJPeH11Eafr34Hxwne/G9+jZCKn9YrpN+qUyvlf7PmABJpnd14T
	 qSzfrj63MTbde5xc6TUvCKbwHMYoNryqbH1k0phw=

**Comment by prakash2jcommerce** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  prakash@twojay.in
    Domain 2jcommerce.in has exceeded the max emails per hour (300/250 (120%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
------ The body of the message is 20053 characters long; only the first
------ 5000 or so are included here.
Received: from out-14.smtp.github.com ([192.30.254.197]:44941)
	by mail.2jcommerce.co.uk with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.87)
	(envelope-from <noreply@github.com>)
	id 1gwojd-000GRA-FR
	for prakash@2jcommerce.in; Thu, 21 Feb 2019 13:49:57 +0000
Date: Thu, 21 Feb 2019 05:49:16 -0800
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1550756956;
	bh=p+G0gFHhJdItGScWKzTLJ30VClc0sAMU8IHf8w9uYyE=;
	h=Date:From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=HoFY8UuQP8QD9Ip4pN49fVN6adRRPq72bfedXS3fFVLxsS2iiWD9NJ0PADjY6Izuj
	 WfP2kAqiBRo8IqPb4yNuuZKtW2rSLO+zdK3+PKDSuC+qjSY462EBSYQILfhHSWTb5j
	 /8Y04Mib5xqC5STx72n9N1xikf9Zrw6GtE7nDcaM=

**Comment by nainesh2jcommerce** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  nainesh@twojay.in
    Domain 2jcommerce.in has exceeded the max emails per hour (305/250 (122%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
------ The body of the message is 20039 characters long; only the first
------ 5000 or so are included here.
Received: from o5.sgmail.github.com ([192.254.113.10]:22209)
	by mail.2jcommerce.co.uk with esmtps (TLSv1.2:ECDHE-RSA-AES128-GCM-SHA256:128)
	(Exim 4.87)
	(envelope-from <bounces+848413-e096-nainesh=2jcommerce.in@sgmail.github.com>)
	id 1gwoje-000GRG-OE
	for nainesh@2jcommerce.in; Thu, 21 Feb 2019 13:49:59 +0000
DKIM-Signature: v=1; a=rsa-sha1; c=relaxed/relaxed; d=github.com; 
	h=from:reply-to:to:cc:in-reply-to:references:subject:mime-version:content-type:content-transfer-encoding:list-id:list-archive:list-post:list-unsubscribe; 
	s=s20150108; bh=Z17zpJjNw/lUXqQwUsCjb/SMyZE=; b=UAL84ezySQLVH1up
	4BrnOOcHo49pfBCX06QEOg7YMSzfzh3synmQQGcUum2g9zHn3JGSyBEK1TLz6MEY
	eN1/354qRtef4Bn+azKRtM4vCvXyP/WbDWEkiRuWDNRPcyXNIyMZaHG26i06YkZT
	yjHBtKF3f4gbjRKN1O6zCQ8o8UU=
Received: by filter0127p1iad2.sendgrid.net with SMTP id filter0127p1iad2-27048-5C6EAC5C-14
        2019-02-21 13:49:16.49611322 +0000 UTC m=+68596.596645115
Received: from github-lowworker-1f7e42f.cp1-iad.github.net (unknown [192.30.252.46])
	by ismtpd0015p1iad2.sendgrid.net (SG) with ESMTP id wELYP2nsQgOrUI58Cn36Qw
	for <nainesh@2jcommerce.in>; Thu, 21 Feb 2019 13:49:16.455 +0000 (UTC)
Received: from github.com (localhost [127.0.0.1])
	by github-lowworker-1f7e42f.cp1-iad.github.net (Postfix) with ESMTP id 4E9B1C0307
	for <nainesh@2jcommerce.in>; Thu, 21 Feb 2019 05:49:16 -0800 (PST)
Date: Thu, 21 Feb 2019 13:49:16 +0000 (UTC)

**Comment by ajay2jcommerce** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  ajay@twojay.in
    Domain 2jcommerce.in has exceeded the max emails per hour (301/250 (120%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
------ The body of the message is 20056 characters long; only the first
------ 5000 or so are included here.
Received: from out-11.smtp.github.com ([192.30.254.194]:54323)
	by mail.2jcommerce.co.uk with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.87)
	(envelope-from <noreply@github.com>)
	id 1gwoje-000GRH-0b
	for ajay@2jcommerce.in; Thu, 21 Feb 2019 13:49:58 +0000
Date: Thu, 21 Feb 2019 05:49:16 -0800
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1550756956;
	bh=0FgZVyyDqWGcZuTpdIlFiWRMr7/RX4UvXOxEJOLEaMU=;
	h=Date:From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=wEEutFh4Es6s6/H/Iy4doOckBcvDdVMu3lkvAlV5+dR91mVn7LXK06d8UzX3pLD0i
	 PFoAi8z2v7RLY7GprNPxMEou5GdwvV9oCmbC4F3VPbcfZrsHeIaHkNgP/cFZYikB/b
	 eUpIGeWc/JD3YP7MemCQhL1euW1YDWDj0IQLBoKw=

**Comment by tomekjordan** [comment]
I can confirm that Magento 2.3.1 creates crazy amounts of duplicated cron tasks, some of them are finished, many are missed, due to duplication. Especially some indexers were not refreshed in Update by Schedule mode. Definitely it should be fixed asap.

**Comment by chickenland** [comment]
The issue here (that I can re-create) is the cron running multiple times in the same minute will schedule for that minute over and over - because it completes the task and then reschedules it.  The fix I envisage here would be that instead of checking for pending cron tasks, it checks in all non-exited statuses (ie. pending, running or success) - and schedules based on that.  Therefore, overlapping, completing/running cron tasks will not re-schedule completing/running cron tasks.

**Comment by magento-engcom-team** [comment]
Hi @udovicic. Thank you for your report. 
The issue has been fixed in magento/magento2#23312 by @chickenland in 2.3-develop branch
Related commit(s):
 - [7859a2fafe944d0001f663b8502143202dea0216](https://github.com/magento/magento2/commit/7859a2fafe944d0001f663b8502143202dea0216)
 - [dd5af3d8daf095642839ab0af3e716e3e6d93be9](https://github.com/magento/magento2/commit/dd5af3d8daf095642839ab0af3e716e3e6d93be9)
 - [8ee5163c3e031522b3fc2091a6703d1bac581297](https://github.com/magento/magento2/commit/8ee5163c3e031522b3fc2091a6703d1bac581297)

The fix will be available with the upcoming 2.3.3 release.

**Comment by magento-engcom-team** [comment]
Hi @udovicic. Thank you for your report. 
The issue has been fixed in magento/magento2#23439 by @ihor-sviziev in 2.2-develop branch
Related commit(s):
 - [e22fb2359efefe91af6f553efdbd934916da415c](https://github.com/magento/magento2/commit/e22fb2359efefe91af6f553efdbd934916da415c)
 - [ff568ac7ceb868cf71a444066be20c9be106b2ba](https://github.com/magento/magento2/commit/ff568ac7ceb868cf71a444066be20c9be106b2ba)
 - [82d3b7161b68ae50459e4439388022c29dd66037](https://github.com/magento/magento2/commit/82d3b7161b68ae50459e4439388022c29dd66037)

The fix will be available with the upcoming 2.2.10 release.

**Comment by sebfie** [comment]
Why this fix is not present on 2.3+ ? It disappears on 2.4.0
